common4j/src/main/com/microsoft/identity/common/java/cache/SharedPreferencesAccountCredentialCache.java

```diff
@@ -100,112 +107,114 @@ public SharedPreferencesAccountCredentialCache(
         Logger.verbose(TAG, "Init: " + TAG);
         mSharedPreferencesFileManager = sharedPreferencesFileManager;
         mCacheValueDelegate = accountCacheValueDelegate;
+        new Thread(() -> load()).start();
+    }
+
+    private void load()
+    {
+        synchronized (mCacheLock) {
+            try {
+                mCachedAccountRecordsWithKeys = loadAccountsWithKeys();
+                Logger.info(TAG, "Loaded " + mCachedAccountRecordsWithKeys.size() + " AccountRecords");
+                mCachedCredentialsWithKeys = loadCredentialsWithKeys();
+                Logger.info(TAG, "Loaded " + mCachedCredentialsWithKeys.size() + " Credentials");
+            } catch (Throwable t) {
+                Logger.error(TAG, "Failed to load initial credentials from SharedPreferences", t);
+            } finally {
+                mLoaded = true;
+                mCacheLock.notifyAll();
+            }
+        }
+    }
+
+    private void waitForInitialLoad() {
+        while (!mLoaded) {
+            try {
+                mCacheLock.wait();
+            } catch (InterruptedException e) {}
+        }
     }
 
     @Override
-    public synchronized void saveAccount(@NonNull final AccountRecord accountToSave) {
+    public void saveAccount(@NonNull final AccountRecord accountInput) {
+        AccountRecord accountToSave = null;
+        try { accountToSave = (AccountRecord) accountInput.clone(); } catch (CloneNotSupportedException e) {}
```

Comment:
If accountInput.clone() fails and throws an exception, we still continue to proceed to saving account with accountToSave which can potentially be null right?

Reply:
Same with other places where we call .clone()

Reply:
These clone()'s actually can't fail.  clone() is a bit weird in Java -- the default Object method defines clone() as protected, and if you want a type to actually be cloneable, you 'implements Cloneable' and then make the clone() method public.  The signature of clone() has the Java checked exception 'CloneNotSupportedException' in it, in case you wind up calling clone() on some object hierarchy that doesn't actually support it.

In this case, our AccountCredentialBase-derived objects do support it, and there's no case where this can fail.

Reply:
I'll add some logging and handling here to handle this, but if we land here, we have hit developer error.

Reply:
If we ever do hit a cloning error, it would be good in that (rare) scenario to return early instead of having an NPE. And yeah logging would help in this case too